### PR TITLE
Change way selector is retrieved in hasFocus

### LINKED ISSUE #2018: hasFocus can't be chained
## The problem

Certain commands do not work when chained from an element request.  (e.g. `hasFocus()`). When invoked as a chained command, they throw the exception 'Cannot read property 'selector' of undefined'. This limits the usefulness of chaining and, in particular, the Pageobject approach.

## Environment

* WebdriverIO version: 4.7.1
* Node.js version: 6.8.0
* [Standalone mode or wdio testrunner](http://webdriver.io/guide/getstarted/modes.html): wdio test runner
* if wdio testrunner, running synchronous or asynchronous tests: synchronous
* Additional wdio packages used (if applicable): ['sauce', 'selenium-standalone', 'static-server']
* Using Mocha test

## Details

The issue appears to be related to the handling of missing parameters.  This line here is the one that throws the exception:  https://github.com/webdriverio/webdriverio/blob/master/lib/commands/hasFocus.js#L30

The `value` of the last promise is `undefined` (and there's no check against that).  Switching to use `this.lastResult.selector` seems to work OK, though I don't know enough about the code base to suggest this change globally.

## Link to Selenium/WebdriverIO logs

https://gist.github.com/webOS101/5058e80082b13e2d071d0e6446cd05c3

## Code To Reproduce Issue [ Good To Have ]

```
		browser.url('http://webdriver.io/');
		$('#docsearch').hasFocus();
```

### PR DESCRIPTION
**Pull request by webOS101**
## Proposed changes

This resolves #2018 by changing the method of retrieving the previous element(s).

## Types of changes

What types of changes does your code introduce to WebdriverIO?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

Could not find any way to reproduce the issue with the current test suite so no test is attached.

### CODE REVIEW

File: lib/commands/hasFocus.js [hunk 1]
```diff
@@ -26,23 +26,34 @@
...
+            var focused = document.activeElement
+            if (!focused) {
+                return false
             }
-        }
-        return instanceOfSelectorHasFocus
-    }, finalSelector)
+
+            var instanceOfSelectorHasFocus = false
+            for (var i = 0; i < selectorArray.length; i++) {
```
Comment (christian-bromann) [inline lib/commands/hasFocus.js]: Shorter:
```js
return selectorArray
    .filter((elem) => focused === elem)
    .filter((hasFocus) => Boolean(hasFocus)
    .length > 0
```

Comment (christian-bromann) [inline lib/commands/hasFocus.js]: Also let's rename `selectorArray` to `elemArray` or something along these lines.

File: lib/commands/hasFocus.js [hunk 2]
```diff
@@ -26,23 +26,34 @@
...
 
+import { RuntimeError } from '../utils/ErrorHandler'
+
 let hasFocus = function (selector) {
-    const finalSelector = selector === null ? this.lastPromise.inspect().value.selector : selector
-    let result = this.execute(function (selector) {
-        var focused = document.activeElement
-        if (!focused) {
-            return false
+    let result = this.elements(selector).then((res) => {
```
Comment (christian-bromann) [inline lib/commands/hasFocus.js]: So this command basically checks if one of the matching elements has focus (so it returns always a boolean value). Let's make sure to note that in the `@return` annotation

Comment (christian-bromann) [inline lib/commands/hasFocus.js]: Let's also let's add `protocol/elements` to the `@uses` annotation

File: test/fixtures/specs/element.spec.js [hunk 3]
```diff
@@ -12,6 +12,11 @@ describe('element as first class citizen', () => {
         expect(header.getLocation('y')).to.be.equal(20)
     })
 
+    it('should be able to use hasFocus on element', () => {
+        let login = browser.element('[name=login]')
+        expect(login.hasFocus()).to.be.true
```
Comment (christian-bromann) [inline test/fixtures/specs/element.spec.js]: The test page doesn't have a element like that afaik

Comment (webOS101) [inline test/fixtures/specs/element.spec.js]: Interesting...  I got that from using `execute` to get the name of `document.activeElement`!

Comment (webOS101) [inline test/fixtures/specs/element.spec.js]: Also, the test passes. :P

File: lib/commands/hasFocus.js (hunk 1 above)
Comment (webOS101) [inline lib/commands/hasFocus.js]: Can do!  Not sure the Boolean filter step is necessary...

**Pull request merged**